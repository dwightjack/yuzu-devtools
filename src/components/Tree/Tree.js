import { html, nothing } from 'lit-html';
import '../Instance/Instance';

export default function Tree({ actions, getData = () => ({}) }) {
  return function renderChild(ids) {
    return ids.map((id) => {
      const {
        Component,
        uid,
        childIds,
        expanded = false,
        selected = false,
        watched = false,
      } = getData(id);
      const { onClick, onSelect } = actions;
      const hasChildren = Array.isArray(childIds) && childIds.length > 0;
      return html`
        <yzdt-component
          name=${Component}
          uid=${uid}
          ?expandable=${hasChildren}
          ?expanded=${expanded}
          ?selected=${selected}
          ?watched=${watched}
          .onClick=${onClick}
          .onSelect=${onSelect}
          >${hasChildren ? renderChild(childIds) : nothing}</yzdt-component
        >
      `;
    });
  };
}
