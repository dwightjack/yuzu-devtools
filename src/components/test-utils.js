import { render, html } from 'lit-html';

export const mockComponent = (tag) => (props) => {
  const el = document.createElement(tag);
  Object.keys(props).forEach((key) => {
    el.setAttribute(key, props[key]);
  });
  return el;
};

export function cycle() {
  return new Promise((resolve) => {
    setTimeout(resolve, 0);
  });
}

const cleanup = (wrapper) =>
  wrapper.querySelectorAll('style').forEach((style) => {
    style.remove();
  });

const toObject = (wrapper) => ({
  $root: wrapper,
  get html() {
    return wrapper.innerHTML;
  },
  find(selector) {
    return wrapper.querySelector(selector);
  },
  findAll(selector) {
    return wrapper.querySelectorAll(selector);
  },
});

export function toHTML(template, props = {}, raw = false) {
  const wrapper = document.createElement('div');

  Object.defineProperty(wrapper, 'isRoot', {
    value: true,
    enumerable: false,
  });

  if (props && typeof template === 'function') {
    // eslint-disable-next-line no-param-reassign
    template = template(props);
  }
  if (Array.isArray(template)) {
    render(
      html`
        ${template}
      `,
      wrapper,
    );
  } else {
    render(template, wrapper);
  }
  cleanup(wrapper);
  return raw ? wrapper : toObject(wrapper);
}

export function toHTMLAsync(template, props, raw = false) {
  const root = toHTML(template, props, true);
  return cycle().then(() => {
    cleanup(root);
    return raw ? root : toObject(root);
  });
}
