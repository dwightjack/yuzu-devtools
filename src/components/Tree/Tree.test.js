import Tree from './Tree';
import { toHTML } from '../test-utils';

describe('Tree', () => {
  let renderer;
  let getData;
  const dummyData = ['_0', '_1'];
  const actions = {
    myAction: () => {},
  };

  const dataMock = (id) => ({ uid: id, Component: `DEMO${id}` });

  beforeEach(() => {
    getData = jest.fn(dataMock);
    renderer = Tree({
      actions,
      getData,
    });
  });

  test('matches default snapshot', () => {
    const { html } = toHTML(renderer(dummyData));
    expect(html).toMatchSnapshot();
  });

  test('matches nested snapshot', () => {
    const childIds = ['_0-1'];
    const spy = jest.fn(renderer);
    getData.mockImplementationOnce((id) => ({
      ...dataMock(id),
      childIds,
    }));
    const { html } = toHTML(spy(dummyData));
    expect(html).toMatchSnapshot();
  });
});
