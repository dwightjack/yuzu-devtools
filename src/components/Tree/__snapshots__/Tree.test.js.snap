// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree matches default snapshot 1`] = `
<yzdt-component name="DEMO_0"
                uid="_0"
>
</yzdt-component>
<yzdt-component name="DEMO_1"
                uid="_1"
>
</yzdt-component>
`;

exports[`Tree matches nested snapshot 1`] = `
<yzdt-component name="DEMO_0"
                uid="_0"
                expandable
>
  <yzdt-component name="DEMO_0-1"
                  uid="_0-1"
  >
  </yzdt-component>
</yzdt-component>
<yzdt-component name="DEMO_1"
                uid="_1"
>
</yzdt-component>
`;
